contrib/raftexample/raft.go

```diff
@@ -247,7 +302,76 @@ func (rc *raftNode) stopHTTP() {
 	<-rc.httpdonec
 }
 
+func (rc *raftNode) publishSnapshot(snapshotToSave raftpb.Snapshot) {
+	if raft.IsEmptySnap(snapshotToSave) {
+		return
+	}
+
+	log.Printf("publishing snapshot at index %d", rc.snapshotIndex)
+	defer log.Printf("finished publishing snapshot at index %d", rc.snapshotIndex)
+
+	if snapshotToSave.Metadata.Index <= rc.appliedIndex {
+		log.Fatalf("snapshot index [%d] should > progress.appliedIndex [%d] + 1", snapshotToSave.Metadata.Index, rc.appliedIndex)
+	}
+
+	spath := filepath.Join(rc.snapdir, fmt.Sprintf("%016x-%016x.snap", snapshotToSave.Metadata.Term, snapshotToSave.Metadata.Index))
+	fpath := filepath.Join(rc.snapdir, "snap")
+	if err := os.Rename(spath, fpath); err != nil {
+		panic(err)
+	}
+	rc.commitC <- nil // trigger kvstore to load snapshot
+
+	rc.confState = snapshotToSave.Metadata.ConfState
+	rc.snapshotIndex = snapshotToSave.Metadata.Index
+	rc.appliedIndex = snapshotToSave.Metadata.Index
+}
+
+func (rc *raftNode) getSnapshot() []byte {
+	buf := new(bytes.Buffer)
+	err := rc.writeSnap(buf)
+	if err != nil {
+		panic(err)
+	}
+	return buf.Bytes()
+}
+
+var snapshotCatchUpEntriesN uint64 = 10000
+
+func (rc *raftNode) maybeTriggerSnapshot() {
+	if rc.appliedIndex-rc.snapshotIndex <= rc.snapCount {
+		return
+	}
+
+	log.Printf("start snapshot [applied index: %d | last snapshot index: %d]", rc.appliedIndex, rc.snapshotIndex)
+	snap, err := rc.raftStorage.CreateSnapshot(rc.appliedIndex, &rc.confState, rc.getSnapshot())
+	if err != nil {
+		panic(err)
+	}
+	if err := rc.saveSnap(snap); err != nil {
+		panic(err)
+	}
+	log.Printf("compacted log at index %d", snap.Metadata.Index)
```

Comment:
from line 345->352, we create a new raft snapshot.

from line 355->361, we compact entries.

So the log should be snapshot at index ..., then compacted log at line ...

Now we have duplicate log lines at line 353 and line 362.